### ISSUE TITLE
Turbo daemon interrupts `git rebase` with git lock file

### ISSUE THREAD

**Issue by dominics** [vercel/turbo#8220]
### Verify canary release

- [X] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

https://github.com/chris-olszewski/turbo-large-monorepo

### What package manager are you using / does the bug impact?

pnpm / npm (haven't tested yarn)

### What operating system are you using?

MacOS 14.4.1

### Which canary version will you have in your reproduction?

2.0.0-canary.0

### Describe the Bug

When the turbo daemon is running, and I perform a rebase of one of my git development branches, the rebase is interrupted repeatedly (every few rebased commits), with a git lockfile error:

```
error: Unable to create '/Users/dominic/workspace/my-project/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
```

If I `turbo daemon stop` the problem no longer occurs.

### Expected Behavior

It seems like it might be unsafe (in terms of breaking users' workflows) for a daemon to lock the `.git/` directory after files have changed, because they'll be changing rapidly and repeatedly during a rebase. I'm not sure if that's something the turbo daemon does, but it seems like it might from having a quick browse of the code?

Perhaps related: I can't see any usage of `--no-optional-locks` nor `GIT_OPTIONAL_LOCKS` within turbo, so even things that look read-only, like `git status`, could cause a lock to be taken (`git status` updates the index as a side-effect).

### To Reproduce

I have a large monorepo (although the branch I'm rebasing is relatively small - 5 commits, maybe 1000 lines, package and file changes mixed):

```
$ du -sh .git/
895M	.git
$ git rev-list --all --count
28050
```

I tried to reproduce the problem on a new repository made with `npx create-turbo@latest -e with-shell-commands`, however in such a repository, `git` operations complete much faster,  and the problem isn't reproducible. I also tried to add additional commits to slow the rebase down, but I couldn't add enough.

When testing, remember to add the `--no-ff` flag to `git rebase` so it actually rebases, if the branch is already up to date.
  
Here is an open source reproduction:

```sh
# obtain a large monorepo
git clone git@github.com:chris-olszewski/turbo-large-monorepo.git
cd turbo-large-monorepo/
npm install # this produces a modified package-lock.json

# commit that change to a new branch so we have a commit to rebase
git checkout -b new
git add package-lock.json
git commit -m 'ran npm install'

# start the daemon
turbo build
# can control-c this after a while, just so `turbo daemon status` shows the daemon is running

# cause the lock failure
git rebase --no-ff main
# repeat this last step (after git rebase --abort) until it fails, but it's a 100% failure rate for me
# turbo daemon stop allows the rebase to complete successfully
```

For canary:

```
$ npm add -DE turbo@canary
$ ./node_modules/.bin/turbo --version
2.0.0-canary.0
$ ./node_modules/.bin/turbo daemon stop
$ ./node_modules/.bin/turbo daemon start
$ git rebase --no-ff main # same failure
```

### Additional context

`turbo daemon logs` contains many lines like:

```
2024-05-28T01:20:27.805686Z  WARN turborepo_lib::package_changes_watcher: file event lagged
2024-05-28T01:20:28.808170Z  WARN turborepo_lib::package_changes_watcher: file event lagged
```

however, it doesn't appear to log any of those at the time the lock occurs.

When I run the daemon with `turbo daemon -vvv`, there are a _lot_ of logs, possibly too many to see what's going on. When I run with slightly less detail (`-vv`), I can see the watcher notice `changed_packages` and `changed_files` (this branch did change the pnpm lockfile). I can also see it receiving events on files from within `.git/`.

But I _can't_ see is why it would be creating its own `index.lock` (or where it does any git operations, or interferes with the lockfile from some other process) - those don't seem to be logged in detail, or I'm not familiar enough with the daemon logging to spot them - I'm happy to email away my daemon logs if it helps.

I'm also unable to say what the `git` process (presumably) spawned by the daemon is doing in any detail, unfortunately: I'm running with SIP (so no good tooling for monitoring git process executions system-wide), and by the time `lsof` displays the process that has produced the lock, it is long gone (so an immediate `ps` doesn't show it, and MacOS `lsof` doesn't show process arguments!). Happy to be pointed to a method that might snarf the full process though.

**Comment by richardhopton** [comment]
We're seeing the same error regularly with our monorepo. I've seen it with rebasing, but also with `lint-staged` which stashes unstaged changes, and pops them afterwards. This can result in changes being lost since `lint-staged` runs as a pre-commit hook.

**Comment by weyert** [comment]
Yes, seeing the same issue when running turbo and trying to commit changes in tools like Sourcetree (Mac)

**Comment by mehulkar** [comment]
https://github.com/vercel/turbo/pull/8247

**Comment by NicholasLYang** [comment]
This should be fixed in 2.0 or 1.13.4. Just restart the daemon with `turbo daemon restart`
